src/codegen/optimizer.lisp

```diff
@@ -316,6 +320,106 @@
      (list
       (action (:after node-application) #'rewrite-application)))))
 
+(defun aggressive-heuristic (node)
+  (and (<= (count-applications node)
+           8)
+       (<= (count-nodes node)
+           32)))
+
+(defun heuristic-inline-applications (node
+                                      env
+                                      &key
+                                        (max-depth  16)
+                                        (max-unroll 4)
+                                        (heuristic  #'aggressive-heuristic))
+  "Recursively inline function calls in `node`, using the environment
+`env` to look up the function bodies to substitute. The traversal
+keeps a call stack and will not continue past the specified
+`max-depth`. For recursive functions, traversal will not continue if
+the function already appears more than `max-unroll` times in the call
+stack. Otherwise, if a function's code can be found in `env` and the
+`heuristic` returns true, then the body will be inlined.
+
+Currently, this may break recursive data let expressions, since the
+codegen requires direct constructor calls."
+  (declare (type node           node)
+           (type tc:environment env)
+           (type (integer 0)    max-depth)
+           (type (integer 0)    max-unroll)
+           (type function       heuristic)
+           (values node &optional))
+  (flet ((try-inline (node call-stack)
+           (declare (type (or node-application node-direct-application) node)
+                    (type parser:identifier-list                        call-stack)
+                    (values (or node-let node-application node-direct-application) &optional))
+           (or
+            (alexandria:when-let*
```

Comment:
It seems `alexandria:if-let` is not an option because only `alexandria:when-let*` has early exiting